### PR TITLE
Feature 1931

### PR DESCRIPTION
**Pull request by RafaFP**
Implemented the "clear public chat" feature. Now moderators can erase the chat history of a conference room's public chat window.

### CODE REVIEW

File: akka-bbb-apps/src/main/java/org/bigbluebutton/core/api/IBigBlueButtonInGW.java [hunk 1]
```diff
@@ -106,6 +106,7 @@ void lockLayout(String meetingID, String setById,
 	void getChatHistory(String meetingID, String requesterID, String replyTo);
 	void sendPublicMessage(String meetingID, String requesterID, Map<String, String> message);
 	void sendPrivateMessage(String meetingID, String requesterID, Map<String, String> message);
+	void clearPublicChatHistory(String meetingID, String requesterID, String replyTo);
```
Comment (fcecagno) [inline akka-bbb-apps/src/main/java/org/bigbluebutton/core/api/IBigBlueButtonInGW.java]: Why do we need a replyTo? Is is used somewhere?

File: bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/client/ChatClientMessageSender.java [hunk 2]
```diff
@@ -111,4 +119,19 @@ private void processGetChatHistoryReply(GetChatHistoryReplyMessage gch) {
 		service.sendMessage(m);
 	}
 
+	private void processClearPublicChatHistoryReply(ClearPublicChatHistoryReplyMessage gcl) {
+
+		Map<String, Object> args = new HashMap<String, Object>();	
+		args.put("meetingId", gcl.meetingId);
+		args.put("requester_id", gcl.requesterId);
+		args.put("message", "ClearPublicChat");
```
Comment (fcecagno) [inline bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/client/ChatClientMessageSender.java]: Do we use this message field?

File: bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/service/ChatService.java [hunk 3]
```diff
@@ -43,6 +43,15 @@ public void sendPublicChatHistory() {
 		red5BBBInGw.getChatHistory(meetingID, requesterID, replyTo);
 	}
 	
+	public void clearPublicChatMessages() {
+		String meetingID = Red5.getConnectionLocal().getScope().getName();
+		String requesterID = getBbbSession().getInternalUserID();
+		// Just hardcode as we don't really need it for flash client. (ralam may 7, 2014)
+		String replyTo = meetingID + "/" + requesterID;
```
Comment (fcecagno) [inline bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/service/ChatService.java]: How did @ralam commented on this code? :)

File: bigbluebutton-client/branding/default/style/css/BBBBlack.css [hunk 4]
```diff
@@ -93,7 +93,7 @@ DataGrid {
 }
 
 
-.whiteboardUndoButtonStyle, .whiteboardCircleButtonStyle, .whiteboardClearButtonStyle,
+.whiteboardUndoButtonStyle, .whiteboardCircleButtonStyle, .clearButtonStyle,
```
Comment (fcecagno) [inline bigbluebutton-client/branding/default/style/css/BBBBlack.css]: Nobody uses the BBBBlack.css, you can revert the changes in this file.

File: bigbluebutton-client/branding/default/style/css/BBBDefault.css [hunk 5]
```diff
@@ -44,10 +44,10 @@ ToolTip {
...
   borderStyle: solid;
   borderThickness: 1;
-  paddingBottom: 3;
-  paddingTop: 3;
-  paddingLeft: 3;
-  paddingRight: 3;
+  paddingBottom: 2;
+  paddingTop: 2;
+  paddingLeft: 2;
+  paddingRight: 2;
```
Comment (fcecagno) [inline bigbluebutton-client/branding/default/style/css/BBBDefault.css:61]: This PR has nothing to do with the whiteboard, why do we change the toolbar style? If we use the same style for the chat toolbar, we should create a new style for it.

File: bigbluebutton-client/branding/default/style/css/BBBDefault.css [hunk 6]
```diff
@@ -306,7 +306,7 @@ DataGrid {
 
 .presentationUploadButtonStyle, .pollStartButtonStyle, .presentationBackButtonStyle, .presentationBackButtonDisabledStyle, .presentationForwardButtonStyle, .presentationForwardButtonDisabledStyle,
 .presentationFitToWidthButtonStyle, .presentationFitToPageButtonStyle, .presentationDownloadButtonStyle,
-.sharedNotesSaveButtonStyle, .sharedNotesNewButtonStyle, .sharedNotesFormatButtonStyle
+.saveButtonStyle, .sharedNotesNewButtonStyle, .sharedNotesFormatButtonStyle, .copyButtonStyle
```
Comment (fcecagno) [inline bigbluebutton-client/branding/default/style/css/BBBDefault.css]: Same thing for the sharedNotesSaveButtonStyle - just create a new style called chatSaveButtonStyle and leave the other styles.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/events/ClearPublicChatEvent.as [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+*
+* You should have received a copy of the GNU Lesser General Public License along
+* with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.
+*
+*/
+package org.bigbluebutton.modules.chat.events
+{
+	import flash.events.Event;
+	
+	import org.bigbluebutton.modules.chat.vo.ChatMessageVO;
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/events/ClearPublicChatEvent.as]: Import isn't used.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/services/ChatMessageService.as [hunk 8]
```diff
@@ -125,11 +129,11 @@ package org.bigbluebutton.modules.chat.services
           moderatorOnlyMsg.toUserID = SPACE;
           moderatorOnlyMsg.toUsername = SPACE;
           moderatorOnlyMsg.message = MeetingModel.getInstance().modOnlyMessage;
-          
+          
           var moderatorOnlyMsgEvent:PublicChatMessageEvent = new PublicChatMessageEvent(PublicChatMessageEvent.PUBLIC_CHAT_MESSAGE_EVENT);
           moderatorOnlyMsgEvent.message = moderatorOnlyMsg;
           moderatorOnlyMsgEvent.history = false;
-          dispatcher.dispatchEvent(moderatorOnlyMsgEvent);
+          dispatcher.dispatchEvent(moderatorOnlyMsgEvent);
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/services/ChatMessageService.as]: Usually we clear this kind of modifications before the PR, and keep only the valuable modifications.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatBox.mxml [hunk 9]
```diff
@@ -764,6 +766,14 @@ with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.
...
 		}
 
+		private function handleClearPublicChatBoxMessages(event:ClearPublicChatEvent):void{
+			if(publicChat){
+				chatMessages.clearPublicChat();
+				invalidateDisplayList();
+        		validateNow();
+        	}
+		}
+
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatBox.mxml:776]: Fix indentation.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatBox.mxml [hunk 10]
```diff
@@ -798,24 +812,27 @@ with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.
...
 	<mx:VBox width="100%" height="100%">
 	  <mx:TextArea id="txtMsgArea" width="100%" height="100%"
-	               styleName="chatControlBarTextMsgStyle"
-				   change="txtMsgAreaChangeHandler(event)"
-	               toolTip="{ResourceUtil.getInstance().getString('bbb.accessibility.chat.chatwindow.input')}"
-			       accessibilityName="{ResourceUtil.getInstance().getString('bbb.chat.input.accessibilityName')}" />
+                 styleName="chatControlBarTextMsgStyle" 
+                 toolTip="{ResourceUtil.getInstance().getString('bbb.accessibility.chat.chatwindow.input')}"
+				 accessibilityName="{ResourceUtil.getInstance().getString('bbb.chat.input.accessibilityName')}" 
+				 />
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatBox.mxml]: Same thing here, if there's no intended modification, you should revert.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml [hunk 11]
```diff
@@ -85,21 +110,36 @@
...
-			   accessibilityName="{ResourceUtil.getInstance().getString('bbb.chat.copyBtn.accessibilityName')}"/>  
+			   accessibilityName="{ResourceUtil.getInstance().getString('bbb.chat.copyBtn.accessibilityName')}"/> 
+	
+	<mx:Button id="clearBtn" 
+			   styleName="clearButtonStyle"
+			   width="22"
+			   height="22"
+			   visible = "{UsersUtil.amIModerator() &amp;&amp; publicChat}"
+			   enabled = "{UsersUtil.amIModerator() &amp;&amp; publicChat}"
+			   includeInLayout = "{UsersUtil.amIModerator() &amp;&amp; publicChat}"
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml]: What's this &amp;&amp; ?

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml [hunk 12]
```diff
@@ -78,6 +90,19 @@
...
+				var clearPupUpWarning:IFlexDisplayObject = PopUpManager.createPopUp(FlexGlobals.topLevelApplication as DisplayObject, ClearPublicChatPopUpWarning, true);
+
+				clearPupUpWarning.x = this.localToGlobal(new Point()).x - clearPupUpWarning.width + clearBtn.width + 7;
+				clearPupUpWarning.y = this.localToGlobal(new Point()).y + clearPupUpWarning.height/2 + clearBtn.height + 5;
+			}
+
+			public function onChangeMyRole(e:ChangeMyRole):void{
+				clearBtn.visible = UsersUtil.amIModerator() && publicChat;
+				clearBtn.enabled = UsersUtil.amIModerator() && publicChat;
+				clearBtn.includeInLayout = UsersUtil.amIModerator() && publicChat;
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml]: You can change it to something like:

`clearBtn.visible = clearBtn.includeInLayout = UsersUtil.amIModerator() && publicChat;`

File: bigbluebutton-client/branding/default/style/css/BBBBlack.css (hunk 4 above)
Comment (RafaFP) [inline bigbluebutton-client/branding/default/style/css/BBBBlack.css]: Reverted.

File: bigbluebutton-client/src/org/bigbluebutton/modules/sharednotes/views/SharedNotesWindow.mxml [hunk 13]
```diff
@@ -346,10 +346,9 @@
...
-			<mx:Button id="btnNew" styleName="sharedNotesNewButtonStyle" width="26" height="26" click="btnNew_clickHandler(event)" toolTip="{ResourceUtil.getInstance().getString('bbb.sharedNotes.new.toolTip')}"/>
+		<mx:HBox width="100%" horizontalAlign="right" paddingTop="0">		<mx:Button id="btnNew" styleName="sharedNotesNewButtonStyle" width="26" height="26" click="btnNew_clickHandler(event)" toolTip="{ResourceUtil.getInstance().getString('bbb.sharedNotes.new.toolTip')}"/>
 			<mx:Button styleName="sharedNotesFormatButtonStyle" width="26" height="26" click="btnToolbar_clickHandler(event)" toolTip="{ResourceUtil.getInstance().getString('bbb.sharedNotes.toolbar.toolTip')}" visible="{options.showToolbarButton}" includeInLayout="{options.showToolbarButton}"/>
-			<mx:Button id="btnSave" styleName="sharedNotesSaveButtonStyle" width="26" height="26" click="btnSave_clickHandler(event)" toolTip="{ResourceUtil.getInstance().getString('bbb.sharedNotes.save.toolTip')}"/>
+			<mx:Button id="btnSave" styleName="saveButtonStyle" width="26" height="26" click="btnSave_clickHandler(event)" toolTip="{ResourceUtil.getInstance().getString('bbb.sharedNotes.save.toolTip')}"/>
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/sharednotes/views/SharedNotesWindow.mxml]: If you revert the changes in the style, this file can be reverted as well.

File: bigbluebutton-client/src/org/bigbluebutton/modules/whiteboard/views/buttons/ClearButton.mxml [hunk 14]
```diff
@@ -20,9 +20,9 @@ with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.
 -->
 <mx:Button xmlns:mx="http://www.adobe.com/2006/mxml"
 		   width="30" height="30"
-		   click="onClick()" styleName="whiteboardClearButtonStyle"
+		   click="onClick()" styleName="clearButtonStyle"
 		   toolTip="{ResourceUtil.getInstance().getString('bbb.highlighter.toolbar.clear')}"
-		   accessibilityName="{ResourceUtil.getInstance().getString('bbb.highlighter.toolbar.clear.accessibilityName')}">
+		   accessibilityName="{ResourceUtil.getInstance().getString('bbb.highlighter.toolbar.clear.accessibilityName')}">
```
Comment (fcecagno) [inline bigbluebutton-client/src/org/bigbluebutton/modules/whiteboard/views/buttons/ClearButton.mxml]: Same thing here.

File: bigbluebutton-client/src/org/bigbluebutton/modules/sharednotes/views/SharedNotesWindow.mxml (hunk 13 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/sharednotes/views/SharedNotesWindow.mxml]: Reverted.

File: bigbluebutton-client/src/org/bigbluebutton/modules/whiteboard/views/buttons/ClearButton.mxml (hunk 14 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/whiteboard/views/buttons/ClearButton.mxml]: Reverted.

File: bigbluebutton-client/branding/default/style/css/BBBDefault.css (hunk 6 above)
Comment (RafaFP) [inline bigbluebutton-client/branding/default/style/css/BBBDefault.css]: Done.

File: bigbluebutton-client/branding/default/style/css/BBBDefault.css (hunk 5 above)
Comment (RafaFP) [inline bigbluebutton-client/branding/default/style/css/BBBDefault.css:61]: Done.

File: record-and-playback/presentation/scripts/publish/presentation.rb [hunk 15]
```diff
@@ -849,7 +849,18 @@ def processChatMessages
...
+						#unless there is a clear event (after the messge was sent) it will be there until the end of the recording 
+						unless $clear_chat_events.last[:timestamp].to_i <= node[:timestamp].to_i
+							$clear_chat_events.reverse_each do |clear|
+								#but if there is a clear after the message, then the first clear after it will be the end time of this message
+								if (clear[:timestamp].to_i >= node[:timestamp].to_i)
+									chat_end = ( translateTimestamp(clear[:timestamp]) / 1000).to_i
+								end
+							end
+						end
+						$xml.chattimeline(:in => chat_start, :out => chat_end, :direction => :down,  :name => chat_sender, :message => chat_message, :target => :chat )
```
Comment (fcecagno) [inline record-and-playback/presentation/scripts/publish/presentation.rb]: I think it can be simplified. Imagine this:
- process clear_chat_events so you have only the timestamps, not the xml nodes
- include $meeting_end at the end of clear_chat_events
- then you can set chat_end as the first element of the sorted clear_chat_events selected to have only the elements with timestamp equal or higher than the chat message timestamp
  It would essentially reduce this code to one line very simple to understand.

File: akka-bbb-apps/src/main/java/org/bigbluebutton/core/api/IBigBlueButtonInGW.java (hunk 1 above)
Comment (RafaFP) [inline akka-bbb-apps/src/main/java/org/bigbluebutton/core/api/IBigBlueButtonInGW.java]: Removed.

File: bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/client/ChatClientMessageSender.java (hunk 2 above)
Comment (RafaFP) [inline bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/client/ChatClientMessageSender.java]: Removed.

File: bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/service/ChatService.java (hunk 3 above)
Comment (RafaFP) [inline bigbluebutton-apps/src/main/java/org/bigbluebutton/red5/service/ChatService.java]: Removed.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatBox.mxml (hunk 9 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatBox.mxml:776]: Done.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml (hunk 12 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml]: Done.

File: bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml (hunk 11 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/chat/views/ChatToolbar.mxml]: I needed both the user to be a moderator and the chat box to be public in order to have this button.

File: bigbluebutton-client/src/org/bigbluebutton/modules/sharednotes/views/SharedNotesWindow.mxml (hunk 13 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/sharednotes/views/SharedNotesWindow.mxml]: Done.

File: bigbluebutton-client/src/org/bigbluebutton/modules/whiteboard/views/buttons/ClearButton.mxml (hunk 14 above)
Comment (RafaFP) [inline bigbluebutton-client/src/org/bigbluebutton/modules/whiteboard/views/buttons/ClearButton.mxml]: Done.
